File: assets/rt/flows/pipelineParallel/PipelineParallelFlowK.java
package rt.flows.pipelineParallel;

import base.Bool_0;
import base.Opt_1;
import base.Void_0;
import base.flows.*;
import rt.FearlessError;

public interface PipelineParallelFlowK extends _PipelineParallelFlow_0 {
  PipelineParallelFlowK $self = new PipelineParallelFlowK(){};

  @Override default Flow_1 fromOp$imm(FlowOp_1 source_m$, Opt_1 size_m$) {
    var source = RestrictFlowReuse_0.$hash$imm$fun(new SafeSource(source_m$), RestrictFlowReuse_0.$self);
    return _PipelineParallelFlow_0.fromOp$imm$fun(source, size_m$, this);
  }
  @Override default Flow_1 $hash$imm(Object e_m$) {
    return _PipelineParallelFlow_0.$hash$imm$fun(e_m$, this);
  }

  /**
   * A variant of FlowOp which catches all FearlessErrors during execution of the operation.
   * We only catch errors in .step/.forRemaining. Any other methods should never throw. This is unenforceable for user-written
   * FlowOps, but we control all FlowOps used within any parallel flows-- so we can make sure to uphold this property.
   * More thought would be needed if we wanted to support parallel flows of user created data structures in the future.
   */
  final class SafeSource implements FlowOp_1 {
    private final FlowOp_1 original;
    private SafeSource(FlowOp_1 original) {
      this.original = original;
    }

    @Override public Bool_0 isFinite$mut() {
      return original.isFinite$mut();
    }

    @Override public Void_0 step$mut(_Sink_1 sink_m$) {
      try {
        return original.step$mut(sink_m$);
      } catch (PipelineParallelFlow.DeterministicFearlessError err) {
        throw err;
      } catch (FearlessError err) {
        sink_m$.pushError$mut(err.info);
        return Void_0.$self;
      } catch (ArithmeticException err) {
        sink_m$.pushError$mut(base.Infos_0.$self.msg$imm(rt.Str.fromJavaStr(err.getMessage())));
        return Void_0.$self;
      }
    }

    @Override public Void_0 stop$mut() {
      return original.stop$mut();
    }

    @Override public Bool_0 isRunning$mut() {
      return original.isRunning$mut();
    }

    @Override public Void_0 forRemaining$mut(_Sink_1 downstream_m$) {
      try {
        return original.forRemaining$mut(downstream_m$);
      } catch (PipelineParallelFlow.DeterministicFearlessError err) {
        throw err;
      } catch (FearlessError err) {
        downstream_m$.pushError$mut(err.info);
        return Void_0.$self;
      } catch (ArithmeticException err) {
        downstream_m$.pushError$mut(base.Infos_0.$self.msg$imm(rt.Str.fromJavaStr(err.getMessage())));
        return Void_0.$self;
      }
    }

    @Override public Opt_1 split$mut() {
      return original.split$mut();
    }

    @Override public Bool_0 canSplit$read() {
      return original.canSplit$read();
    }
  }
}

File: assets/rt-js/flows/pipelineParallel/PipelineParallelFlowK.js
import { base$$Infos_0 } from "../../../base/Infos_0.js";
import { base$$Void_0 } from "../../../base/Void_0.js";
import { base$$flows$$_PipelineParallelFlow_0 } from "../../../base/flows/_PipelineParallelFlow_0.js";
import { PipelineParallelFlow } from "./PipelineParallelFlow.js";
import { FearlessError } from "../../FearlessError.js";

export class PipelineParallelFlowK {
  /**
   * @param {base$$flows$$FlowOp_1} source_m$
   * @param {base$$Opt_1} size_m$
   * @returns {base$$flows$$Flow_1}
   */
  fromOp$imm$2(source_m$, size_m$) {
    const source = new SafeSource(source_m$);
    return base$$flows$$_PipelineParallelFlow_0.fromOp$imm$3$fun(source, size_m$, this);
  }

  /**
   * Delegates to generated helper
   * @param {any} e_m$
   */
  $hash$imm$1(e_m$) {
    return base$$flows$$_PipelineParallelFlow_0.$hash$imm$2$fun(e_m$, this);
  }

  // singleton for compiler/runtime compatibility
  static $self = new PipelineParallelFlowK();
}

/**
 * SafeSource: wraps a FlowOp_1 and catches FearlessError / arithmetic exceptions
 * so they are forwarded to the downstream sink via pushError$mut$1, mirroring the Java behavior.
 */
class SafeSource {
  /**
   * @param {base$$flows$$FlowOp_1} original
   */
  constructor(original) {
    this.original = original;
    /** @type {PipelineParallelFlow.WrappedSink | null} */
    this.sink = null;
  }

  isFinite$mut$0() {
    return this.original.isFinite$mut$0();
  }

  step$mut$1(sink_m$) {
    // keep reference to wrapped sink when observed
    if (sink_m$ instanceof PipelineParallelFlow.WrappedSink) {
      this.sink = sink_m$;
    }
    try {
      return this.original.step$mut$1(sink_m$);
    } catch (err) {
      // rethrow deterministic wrapper errors unchanged
      if (err instanceof PipelineParallelFlow.DeterministicFearlessError) {
        throw err;
      }
      if (err instanceof FearlessError) {
        // forward the Info to the sink as pushError$mut$1
        try {
          sink_m$.pushError$mut$1(err.info);
        } catch (inner) {
          // If pushError itself throws, bubble the original err as deterministic
          throw new PipelineParallelFlow.DeterministicFearlessError(err.info);
        }
        return base$$Void_0.$self;
      }
      // treat certain JS errors as arithmetic-like exceptions
      if (err instanceof Error) {
        const msg = err.message || String(err);
        try {
          if (base$$Infos_0 && base$$Infos_0.$self && typeof base$$Infos_0.$self.msg$imm$1 === "function") {
            sink_m$.pushError$mut$1(base$$Infos_0.$self.msg$imm$1(msg));
          } else {
            // best-effort fallback
            sink_m$.pushError$mut$1(msg);
          }
        } catch (inner) {
          // convert to deterministic if pushError throws FearlessError
          if (inner instanceof FearlessError) {
            throw new PipelineParallelFlow.DeterministicFearlessError(inner.info);
          }
          // otherwise rethrow inner
          throw inner;
        }
        return base$$Void_0.$self;
      }
      // unknown throwables: rethrow
      throw err;
    }
  }

  stopUp$mut$0() {
    if (this.sink != null) {
      try {
        this.sink.softClose();
      } catch (e) {
        // ignore softClose errors; proceed to stop original
      }
    }
    return this.original.stopUp$mut$0();
  }

  isRunning$mut$0() {
    return this.original.isRunning$mut$0();
  }

  for$mut$1(downstream_m$) {
    if (downstream_m$ instanceof PipelineParallelFlow.WrappedSink) {
      this.sink = downstream_m$;
    }
    try {
      return this.original.for$mut$1(downstream_m$);
    } catch (err) {
      if (err instanceof PipelineParallelFlow.DeterministicFearlessError) {
        throw err;
      }
      if (err instanceof FearlessError) {
        try {
          downstream_m$.pushError$mut$1(err.info);
        } catch (inner) {
          throw new PipelineParallelFlow.DeterministicFearlessError(inner.info);
        }
        return base$$Void_0.$self;
      }
      if (err instanceof Error) {
        const msg = err.message || String(err);
        try {
          if (base$$Infos_0 && base$$Infos_0.$self && typeof base$$Infos_0.$self.msg$imm$1 === "function") {
            downstream_m$.pushError$mut$1(base$$Infos_0.$self.msg$imm$1(msg));
          } else {
            downstream_m$.pushError$mut$1(msg);
          }
        } catch (inner) {
          if (inner instanceof FearlessError) {
            throw new PipelineParallelFlow.DeterministicFearlessError(inner.info);
          }
          throw inner;
        }
        return base$$Void_0.$self;
      }
      throw err;
    }
  }

  split$mut$0() {
    return this.original.split$mut$0();
  }

  canSplit$read$0() {
    return this.original.canSplit$read$0();
  }
}
